File: java-completed/app/src/main/java/com/vonage/tutorial/voice/view/incommingcall/IncomingCallFragment.java
package com.vonage.tutorial.voice.view.incommingcall;

import android.os.Bundle;
import android.view.View;
import android.widget.TextView;
import android.widget.Toast;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import com.google.android.material.floatingactionbutton.FloatingActionButton;
import com.vonage.tutorial.R;
import com.vonage.tutorial.voice.BackPressHandler;

public class IncomingCallFragment extends Fragment implements BackPressHandler {

    private IncomingCallViewModel viewModel;

    FloatingActionButton hangupFab;

    FloatingActionButton answerFab;

    TextView otherUserNameTextView;

    public IncomingCallFragment() {
        super(R.layout.fragment_incoming_call);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        viewModel = new ViewModelProvider(requireActivity()).get(IncomingCallViewModel.class);

        hangupFab = view.findViewById(R.id.hangupFab);
        answerFab = view.findViewById(R.id.answerFab);
        otherUserNameTextView = view.findViewById(R.id.otherUserNameTextView);

        IncomingCallFragmentArgs args = IncomingCallFragmentArgs.fromBundle(getArguments());
        viewModel.onInit(args);

        viewModel.toastLiveData.observe(getViewLifecycleOwner(), it -> Toast.makeText(requireActivity(), it, Toast.LENGTH_SHORT));
        viewModel.otherUserNameLiveData.observe(getViewLifecycleOwner(), it -> otherUserNameTextView.setText(it));

        hangupFab.setOnClickListener(it -> viewModel.hangup());

        answerFab.setOnClickListener(it -> viewModel.answer());
    }


    @Override
    public void onBackPressed() {
        viewModel.onBackPressed();
    }
}

File: kotlin-completed/app/src/main/java/com/vonage/tutorial/voice/view/incommingcall/IncomingCallFragment.kt
package com.vonage.tutorial.voice.view.incommingcall

import android.os.Bundle
import android.view.View
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.lifecycle.Observer
import androidx.navigation.fragment.navArgs
import com.vonage.tutorial.R
import com.vonage.tutorial.voice.BackPressHandler
import com.vonage.tutorial.voice.extension.observe
import com.vonage.tutorial.voice.extension.toast
import kotlinx.android.synthetic.main.fragment_incoming_call.*

class IncomingCallFragment : Fragment(R.layout.fragment_incoming_call),
    BackPressHandler {

    private val toastObserver = Observer<String> {
        context?.toast(it)
    }

    private val otherUserNameLiveData = Observer<String> {
        otherUserNameTextView.text = it
    }

    private val viewModel by viewModels<IncomingViewModel>()
    private val args by navArgs<IncomingCallFragmentArgs>()

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        viewModel.onInit(args)

        observe(viewModel.toastLiveData, toastObserver)
        observe(viewModel.otherUserNameLiveData, otherUserNameLiveData)

        hangupFab.setOnClickListener {
            viewModel.hangup()
        }

        answerFab.setOnClickListener {
            viewModel.answer()
        }
    }

    override fun onBackPressed() {
        viewModel.onBackPressed()
    }
}
